### PR TITLE
Fix handling of defaults on object child columns

### PR DESCRIPTION
**Pull request by mfussenegger**
See https://github.com/crate/crate/pull/14190#discussion_r1203465852

5.3.0 unintentionally changed the behavior.

Closes https://github.com/crate/crate/issues/14189

### CODE REVIEW

File: server/src/test/java/io/crate/integrationtests/ObjectColumnTest.java [hunk 1]
```diff
@@ -343,4 +348,30 @@ public void testSelectUnknownObjectColumnPreservesTheUnknownName() throws Except
...
+                        key text default 'nested default'
+                    ))
+                )
+            )
+            """
+        );
+        assertThat(execute("insert into tbl (x) values (1)")).hasRowCount(1);
+        execute("refresh table tbl");
+        assertThat(execute("select o, os, complex from tbl")).hasRows(
+            "{key=default-value}| [{key=default-value}]| {os=[{key=nested default}]}"
```
Comment (mfussenegger) [inline server/src/test/java/io/crate/integrationtests/ObjectColumnTest.java:374]: Thinking about this some more I'm not sure if adding a single list item is good. I guess arrays should just be empty?

In 5.2 the case led to an error:

```
cr> create table tbl (
        x int,
        o object as (
            key text default 'default-value'
        ),
        os array(object as (
            key text default 'default-value'
        )),
        complex object as (
            os array(object as (
                key text default 'nested default'
            ))
        )
    );
CREATE OK, 1 row affected  (0.102 sec)

cr> insert into tbl (x) values (1);
INSERT OK, 1 row affected  (0.032 sec)

cr> select * from tbl;
ClassCastException[class java.util.HashMap cannot be cast to class [Ljava.lang.Object; (java.util.HashMap and [Ljava.lang.Object; are in module java.base of loader 'bootstrap')]
```

**Review by matriv — APPROVED** [review]
Thank you!

**Comment by BaurzhanSakhariev** [comment]
I think we can also re-enable `IndexerTest#test_default_for_full_object` now?

File: server/src/test/java/io/crate/integrationtests/ObjectColumnTest.java (hunk 1 above)
Comment (matriv) [reply server/src/test/java/io/crate/integrationtests/ObjectColumnTest.java:374]: hm true, maybe forbid completely to have defaults for objects within an array?

**Comment by mfussenegger** [comment]
> I think we can also re-enable `IndexerTest#test_default_for_full_object` now?

No, the analyzer restriction if `DEFAULT` is placed on the object column is still there.

File: server/src/test/java/io/crate/integrationtests/ObjectColumnTest.java (hunk 1 above)
Comment (mfussenegger) [reply server/src/test/java/io/crate/integrationtests/ObjectColumnTest.java]: Pushed a fixup to not create arrays

**Review by matriv — APPROVED** [review]
Thank you, left a comment to improve the note.
Also, do we really want to close the issue with default on the whole object? imho it's still valid to provide a full object if it's null.

**Comment by mfussenegger** [comment]
> Also, do we really want to close the issue with default on the whole object? imho it's still valid to provide a full object if it's null.

Yes we do, if it is important it will pop up again - hopefully with a concrete use case.

**Pull request merged**